### PR TITLE
Improve HTTP Auto-Documentation

### PR DESCRIPTION
**Pull request by HonakerM**
**What this PR does / why we need it**:
This PR improves the swagger ui auto-documentation by adding the following functionalities:
- [ ] Task, DataObject, Module docstrings in HTTP docs
- [ ] Non-flattened OpenAPI Schema
- [ ] Hide-able tasks
- [ ] Customizable Task and DataObject field openapi schemas

**Special notes for your reviewer**:
This PR depends on the [better defaults PR](https://github.com/caikit/caikit/pull/726)

**If applicable**:
- [ ] this PR contains documentation
- [ ] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

### CODE REVIEW

**Comment by gabe-l-hart** [comment]
@HonakerM I'm going to hold on this review until you take it out of Draft or ping me!

**Comment by HonakerM** [comment]
Will do! This PR depends on https://github.com/caikit/caikit/pull/726 so it'll be after that one

File: caikit/runtime/http_server/utils.py [hunk 1]
```diff
@@ -16,31 +16,42 @@
...
+) -> (dict, dict):
     """Helper function to extract all "bytes" or File fields from a json schema and return the
     cleaned schema dict and a dict of extracted schemas where the key is the original raw's path"""
     if isinstance(json_schema, dict):
         # If this json_schema represents a raw field extract it
         if raw_json_schema := _parse_raw_json_schema(json_schema):
             return None, {_clean_schema_path(current_path): raw_json_schema}
 
+        # If this json_schema is just a ref then just recurse on the ref
+        if "$ref" in json_schema:
```
Comment (gabe-l-hart) [inline caikit/runtime/http_server/utils.py:47]: Im not really following what's happening here, but I think it looks correct ish?

**Review by gabe-l-hart — CHANGES_REQUESTED** [review]
This looks like a really nice improvement. The one part that I think needs more discussion is the task metadata and whether we can keep knowledge of `openapi` out of the core cor now.

File: caikit/runtime/http_server/http_server.py [hunk 2]
```diff
@@ -1004,6 +1039,59 @@ def _tls_files(self) -> Iterable[_TlsFiles]:
...
+
+        """
+        # Parse the library name into a more human readable version
+        library_name = "FastAPI"
+        if get_config().runtime.library:
+            library_name = snake_to_upper_camel(get_config().runtime.library)
+
+        # Attempt to load in the runtime library to fetch the module's docstring
+        try:
+            imported_module = importlib.import_module(get_config().runtime.library)
```
[the remark this answers is not in the document]
Comment (HonakerM) [reply caikit/runtime/http_server/http_server.py:1057]: > Is there any chance that this will change import-time semantics?

Hmmm, I actually didn't think about that. I think it should be fine since starting the runtime server happens after task&module discovery.  I liked the idea of having the module docstr in the openapi spec but maybe it should be a config variable instead?

File: caikit/core/task.py [hunk 3]
```diff
@@ -241,6 +243,20 @@ def get_output_type(cls, output_streaming: bool) -> Type[DataBase]:
...
 
+    @classmethod
+    def get_visibility(cls) -> bool:
+        """Get the visibility for this task.
+
+        NOTE: defaults to True even if visibility wasn't provided"""
+        return cls.__annotations__.get(_VISIBLE_ANNOTATION, True)
+
+    @classmethod
+    def get_extra_openapi_schema(cls) -> Dict[str, Any]:
```
[the remark this answers is not in the document]
Comment (HonakerM) [reply caikit/core/task.py:254]: >  I wonder if it would make more sense to rephrase this as a more generic metadata property that could be interpreted as openapi in the http_server?

I think that makes sense, and we could define a `openapi_schema` key in the HTTP server to keep metadata generic

File: caikit/core/task.py [hunk 4]
```diff
@@ -359,6 +377,12 @@ def run_bidi_stream(raw_documents: DataStream[caikit.interfaces.nlp.RawDocument]
             task, which all modules' streaming-output inference methods must return. This must be
             in the form Iterable[T].
 
+        visible (bool): If this task should be exposed to the end user in documentation or if
+          it should only be used internally
+
+        extra_openapi_schema (Optional[Dict[str, Any]]): Any additional openapi schema that should
+          be included in the documentation for this task
```
Comment (gabe-l-hart) [inline caikit/core/task.py:384]: NIT to GitHub! The "suggestion" feature needs the ability to set a word-wrap. Probably need to verify this

```suggestion
        metadata (Optional[Dict[str, Any]]): Any additional interface
            information (e.g. OpenAPI schema) that should be included in the
            documentation for this task
```

File: caikit/runtime/http_server/http_server.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (gabe-l-hart) [reply caikit/runtime/http_server/http_server.py:1058]: I think that module discovery actually _doesn't_ happen after booting the runtime, but as part of it (I had to think about this for a while). The path is:

1. The server base asks for one or more service packages ([here](https://github.com/caikit/caikit/blob/main/caikit/runtime/server_base.py#L85))
2. The `ServicePackageFactory` generates the data model ([here](https://github.com/caikit/caikit/blob/main/caikit/runtime/service_factory.py#L160))
3. The `get_data_model` function calls `_get_cdm_from_lib` ([here](https://github.com/caikit/caikit/blob/main/caikit/runtime/utils/import_util.py#L103)) which calls `get_dynamic_module` ([here](https://github.com/caikit/caikit/blob/main/caikit/runtime/utils/import_util.py#L143))
4. `get_dynamic_module` does the import ([here](https://github.com/caikit/caikit/blob/main/caikit/runtime/utils/import_util.py#L179))

Given all that, I can comfortably say that this is fine since it is invoked in `__init__` after `super().__init__()` which is what triggers the import. I think this tangled story probably belongs in a comment though!

Also, it _might_ be worth using `caikit.runtime.utils.import_utils.get_dynamic_module` here so that we consolidate all questions/concerns about `importlib` over there.

**Review by gabe-l-hart — CHANGES_REQUESTED** [review]
One NIT-pick comment change request, and some fun spelunking on how our import time stuff happens that could use a comment at the minimum.

File: caikit/runtime/http_server/http_server.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (HonakerM) [reply caikit/runtime/http_server/http_server.py:1058]: That makes sense and matches my surface level understanding. I'll throw this in a comment in `http_server` so people know not to move it.

> Also, it might be worth using `caikit.runtime.utils.import_utils.get_dynamic_module`

Definitely! I'll use that instead

**Review by gabe-l-hart — APPROVED** [review]
Ship it!
